### ISSUE TITLE
Creating groupNames from Maven groupId

### ISSUE THREAD

**Issue by JordanJLopez** [renovatebot/renovate#4673]
**What would you like Renovate to be able to do?**
As a Java developer whose apps usually have a very large number of dependencies, it would be great if there was a way to automatically group up dependencies based on the Maven `groupId`.

**Describe the solution you'd like**
I notice that, on the Javascript side of things, there's already a number of Renovate presets that have done this kind of tedious work already (like the [presets-packages](https://docs.renovatebot.com/presets-packages/)). I feel that Java/Maven dependencies actually have a leg up in this regard as they follow the predictable pattern of `${groupId}:${artifactId}`.

It'd be great if there was an option to automatically group maven dependencies by `groupId`

**Describe alternatives you've considered**
The current way to achieve this is with `packageRules` using `packagePatterns`.
This works well for one or two or five, but when I'm pulling in 30 or 40 dependencies, it feels tedious to go and manually define the groupName for each one.

**Additional context**
Currently using Renovate Pro v0.17.0-pre.1 on Enterprise GitHub

**Comment by rarkins** [comment]
@zharinov this sounds like a good idea. The question is then how to manage any conflicts with our existing placeholder grouping. i.e. if 5 dependencies share the same groupId but 4 of them use a shared placeholder version, should we be grouping 4 of them together or 5 of them?

**Comment by zharinov** [comment]
@rarkins Yes, the way we're handling placeholdered versions is a bit hacky right now. I think if we're introducing other grouping strategies, then they should have priority over just placeholders. Also it may lead to reconsidering best strategy for choosing right version: first, last, highest or lowest common.

Hi @JordanJLopez, what do you think about going even further and group by `artifactId` starting parts, [like this](https://github.com/zharinov/java-test-10/pull/4)? My point here is to consider separation of things like `org.example:simple-utils` from packages like `org.example:framework-core`, `org.example:framework-bricks`, `org.example:framework-sticks`, etc.

**Comment by rarkins** [comment]
It sounds like for Maven or java in general we need to formalise our different behaviours and:
- Pick the most likely one automatically
- Allow users to override by config

Approaches include:
- Grouping by groupId
- Grouping by groupId + artifact prefix
- Grouping by placeholder version
- Grouping by source repository

Challenges:
- If users employ a placeholder version then we kind of must include them all in the same PR/group, otherwise you get separate PRs updating the same placeholder
- Sometimes a placeholder version is used by a list of artifacts but they don't all share a common latest release. In my opinion this is a sign of *incorrect* placeholder version use and such users should probably stop using it and instead use groupId grouping and let Renovate update them individually.

**Comment by rarkins** [comment]
As a first step, maybe we can capture `groupId` as part of our extract step and then let people configure it as part of the branch template instead of using `depName` (which incorporates both groupId and artifactId).

**Comment by rarkins** [comment]
@JordanJLopez How often would it be the case that you *don't* want PRs grouped by groupId?

**Comment by JordanJLopez** [comment]
> Hi @JordanJLopez, what do you think about going even further and group by `artifactId` starting parts, [like this](https://github.com/zharinov/java-test-10/pull/4)? My point here is to consider separation of things like `org.example:simple-utils` from packages like `org.example:framework-core`, `org.example:framework-bricks`, `org.example:framework-sticks`, etc.

@zharinov I think that would be ideal, but I'm not sure how one would achieve that kind of matching without specifying the starting parts manually. If there was a good way to do that though, it'd be awesome.

> @JordanJLopez How often would it be the case that you don't want PRs grouped by groupId?

Let me expand a bit on the use case at the company I work at.

Our groupIds usually follow the pattern of
`com.${companyName}.${teamName}` and the artifactIds are _usually_ of the form `${libraryName}-${module}`.

So the groupId matching would basically translate to team matching (for internal dependencies).

If groupId matching was a reality, I can foresee there would be cases where, for specific libraries or frameworks by a given team, we'd want them to have their own individual rules. But generally, grouping by the groupId would usually be preferred.

**Comment by rarkins** [comment]
Here is where the defaults for branch construction are defined: https://github.com/renovatebot/renovate/blob/3eede6ad472452c6805435a84510f1a21b4f0bfe/lib/config/definitions.ts#L1113

Maven inherits from this. Right now `depNameSanitized` comes from the full group + artifact IDs, which means Maven PRs default to one artifact per PR. If we extract `groupName` and generate `groupNameSanitized` then we could easily let people change to using that in the branch, and grouping would follow. We need to make sure we use names that don't contradict/confuse with our existing "groupName" concept though, which is different. Generically it could be "depFamilyName" perhaps?

**Comment by zharinov** [comment]
By the way, here is related question: https://github.com/renovatebot/config-help/issues/378#issuecomment-541028520

**Comment by zharinov** [comment]
Here is [basic solution](https://github.com/renovatebot/renovate/pull/4659/files) which produces outputs [like this](https://github.com/zharinov/java-test-11/pulls) being run against [custom-configured templates](https://github.com/zharinov/java-test-11/blob/master/renovate.json).

It separates placeholder-based PRs from those based on `groupId` or artifact prefixes. If this is good enough, then we need add template examples to README and maybe make it kind of presets.

**Comment by irgaly** [comment]
I'm using Google Play Services Libraries in our Android apps.

https://developers.google.com/android/guides/setup

Google Play Services Libraries package names are:

* com.google.android.gms:play-services-location
* com.google.android.gms:play-services-maps
* com.google.android.gms:play-services-auth
* ...

They have independent versions.

Then, I configured this way for creating separated PRs for each dependency.

```
  "packageRules": [
...
      {
      "groupName": "com.google.android.gms:play-services-location",
      "matchPackageNames": [
        "com.google.android.gms:play-services-location"
      ]
    },
    {
      "groupName": "com.google.android.gms:play-services-maps",
      "matchPackageNames": [
        "com.google.android.gms:play-services-maps"
      ]
    },
    {
      "groupName": "com.google.android.gms:play-services-auth",
      "matchPackageNames": [
        "com.google.android.gms:play-services-auth"
      ]
    }
  ]
```

Are there any better configurations than this?

**Comment by rarkins** [comment]
@irgaly please start a new discussion if you're after config help. Please also describe when you're doing that: what leads you to need config in the first place?

**Comment by irgaly** [comment]
Sorry, I'm toking about this issue's discussion. I'd like to show `com.google.android.gms` is a worth case example of taking groupName from Maven groupId.

Google Services Libraries have many functionality and they are independent each other, so It should not grouping together by groupName `com.google.android.gms`.

![image](https://user-images.githubusercontent.com/1311446/152923024-101fde02-00d1-4cb6-a677-69ebdef82d22.png)

Taking groupName from Maven groupId is a nice logic, and it works fine in many cases. It will be nice more if we can configure to disable or change the Maven groupId grouping rule.

For example, I'm happy if I can configure "don't group `com.google.android.gms`" with like below:

```
  "packageRules": [
...
      {
      "disableMavenGrouping": true,
      "matchPackagePrefixes": [
        "com.google.android.gms:"
      ]
    }
```

or other idea:

```
  "packageRules": [
...
      {
      "groupName": "${mavenGroupId}:${mavenArtifactId}",
      "matchPackagePrefixes": [
        "com.google.android.gms:"
      ]
    }
```
